Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/auto/AutoSubmission.kt

```diff
@@ -0,0 +1,144 @@
+import de.rki.coronawarnapp.submission.SubmissionSettings
+import de.rki.coronawarnapp.submission.task.SubmissionTask
+import de.rki.coronawarnapp.task.TaskController
+import de.rki.coronawarnapp.task.TaskInfo
+import de.rki.coronawarnapp.task.common.DefaultTaskRequest
+import de.rki.coronawarnapp.task.submitBlocking
+import de.rki.coronawarnapp.util.TimeStamper
+import de.rki.coronawarnapp.worker.BackgroundConstants
+import kotlinx.coroutines.flow.map
+import org.joda.time.Instant
+import timber.log.Timber
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class AutoSubmission @Inject constructor(
+    private val timeStamper: TimeStamper,
+    private val submissionSettings: SubmissionSettings,
+    private val workManager: WorkManager,
+    private val taskController: TaskController
+) {
+
+    private var currentMode: Mode = Mode.DISABLED
+
+    private fun List<TaskInfo>.isSubmissionTaskRunning() = any {
+        it.taskState.isActive && it.taskState.request.type == SubmissionTask::class
+    }
+
+    val isSubmissionRunning = taskController.tasks.map { it.isSubmissionTaskRunning() }
+
+    fun setup() {
+        Timber.tag(TAG).v("setup()")
+
+        if (submissionSettings.autoSubmissionEnabled.value) {
+            Timber.tag(TAG).i("Fresh app start and auto submission is enabled, updating mode.")
+
+            updateMode(Mode.SUBMIT_ASAP)
+        } else {
+            Timber.tag(TAG).d("AutoSubmission is disabled.")
+        }
+    }
+
+    fun updateMode(newMode: Mode) {
+        Timber.tag(TAG).i("updateMode(mode=$newMode)")
+
+        when (newMode) {
+            Mode.DISABLED -> disableAutoSubmission()
+            Mode.MONITOR -> enableAutoSubmission(lastActivity = timeStamper.nowUTC)
+            Mode.SUBMIT_ASAP -> enableAutoSubmission(lastActivity = Instant.EPOCH)
+        }
+        currentMode = newMode
+    }
+
+    suspend fun runSubmissionNow() {
+        Timber.tag(TAG).i("runSubmissionNow()")
+
+        val result = taskController.submitBlocking(
+            DefaultTaskRequest(
+                type = SubmissionTask::class,
+                arguments = SubmissionTask.Arguments(checkUserActivity = false),
+                originTag = TAG
+            )
+        )
+        if (result.isSuccessful) {
+            Timber.tag(TAG).i("Blocking submission was successful.")
+        } else {
+            Timber.tag(TAG).w(result.error, "Blocking submission was not successful, enabling auto submission.")
+            updateMode(Mode.SUBMIT_ASAP)
+        }
+    }
+
+    private fun scheduleWorker() {
+        Timber.tag(TAG).v("scheduleWorker(): Creating periodic worker request for submission.")
+
+        val request = PeriodicWorkRequestBuilder<SubmissionWorker>(15, TimeUnit.MINUTES).apply {
+            addTag(AUTOSUBMISSIO_WORKER_TAG)
+            setBackoffCriteria(BackoffPolicy.EXPONENTIAL, BackgroundConstants.BACKOFF_INITIAL_DELAY, TimeUnit.MINUTES)
```

Comment:
short question:
setting a BackOffPolicy here, does it have any effect? (worker doesn't use Result.retry() or something)
Not sure about that...

Reply:
I've changed the worker to return `retry()` on exceptions.